File: enode/src/main/java/org/enodeframework/infrastructure/impl/AbstractHandlerProvider.java
package org.enodeframework.infrastructure.impl;

import org.enodeframework.annotation.Command;
import org.enodeframework.annotation.Event;
import org.enodeframework.annotation.Priority;
import org.enodeframework.annotation.Subscribe;
import org.enodeframework.common.container.IObjectContainer;
import org.enodeframework.common.exception.EnodeRuntimeException;
import org.enodeframework.common.exception.RegisterComponentException;
import org.enodeframework.infrastructure.IAssemblyInitializer;
import org.enodeframework.infrastructure.IObjectProxy;
import org.enodeframework.infrastructure.MethodInvocation;
import org.enodeframework.messaging.MessageHandlerData;
import org.reflections.ReflectionUtils;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public abstract class AbstractHandlerProvider<TKey, THandlerProxyInterface extends IObjectProxy & MethodInvocation, THandlerSource> implements IAssemblyInitializer {
    private final Map<TKey, List<THandlerProxyInterface>> handlerDict = new HashMap<>();
    private final Map<TKey, MessageHandlerData<THandlerProxyInterface>> messageHandlerDict = new HashMap<>();
    private final MethodHandles.Lookup lookup = MethodHandles.lookup();

    protected abstract TKey getKey(Method method);

    protected abstract Class<? extends THandlerProxyInterface> getHandlerProxyImplementationType();

    protected abstract boolean isHandlerSourceMatchKey(THandlerSource handlerSource, TKey key);

    protected abstract boolean isHandleMethodMatch(Method method);

    protected abstract IObjectContainer getObjectContainer();

    @Override
    public void initialize(Set<Class<?>> componentTypes) {
        componentTypes.stream().filter(this::isHandlerType).forEach(this::registerHandler);
        initializeHandlerPriority();
    }

    public List<MessageHandlerData<THandlerProxyInterface>> getHandlers(THandlerSource source) {
        List<MessageHandlerData<THandlerProxyInterface>> handlerDataList = new ArrayList<>();
        messageHandlerDict.keySet().stream()
                .filter(key -> isHandlerSourceMatchKey(source, key))
                .forEach(key -> handlerDataList.add(messageHandlerDict.get(key)));
        return handlerDataList;
    }

    private void initializeHandlerPriority() {
        handlerDict.forEach((key, handlers) -> {
            MessageHandlerData<THandlerProxyInterface> handlerData = new MessageHandlerData<>();
            List<THandlerProxyInterface> listHandlers = new ArrayList<>();
            Map<THandlerProxyInterface, Integer> queueHandlerDict = new HashMap<>();
            handlers.forEach(handler -> {
                int priority = getHandleMethodPriority(handler);
                if (priority == 0) {
                    listHandlers.add(handler);
                } else {
                    queueHandlerDict.put(handler, priority);
                }
            });
            handlerData.allHandlers = handlers;
            handlerData.listHandlers = listHandlers;
            handlerData.queuedHandlers = queueHandlerDict.entrySet().stream().sorted(Comparator.comparingInt(Map.Entry::getValue)).map(x -> x.getKey()).collect(Collectors.toList());
            messageHandlerDict.put(key, handlerData);
        });
    }

    private int getHandleMethodPriority(THandlerProxyInterface handler) {
        Method method = handler.getMethod();
        int priority = 0;
        Priority methodPriority = method.getAnnotation(Priority.class);
        if (methodPriority != null) {
            priority = methodPriority.value();
        }
        if (priority == 0) {
            Priority classPriority = handler.getInnerObject().getClass().getAnnotation(Priority.class);
            if (classPriority != null) {
                priority = classPriority.value();
            }
        }
        return priority;
    }

    private boolean isHandlerType(Class<?> type) {
        if (type == null) {
            return false;
        }
        if (type.isInterface()) {
            return false;
        }
        if (Modifier.isAbstract(type.getModifiers())) {
            return false;
        }
        return containsHandleType(type);
    }

    private boolean containsHandleType(Class<?> type) {
        return type.isAnnotationPresent(Command.class) || type.isAnnotationPresent(Event.class);
    }

    protected boolean isMethodAnnotationSubscribe(Method method) {
        return method.isAnnotationPresent(Subscribe.class);
    }

    private void registerHandler(Class<?> handlerType) {
        Set<Method> handleMethods = ReflectionUtils.getMethods(handlerType, this::isHandleMethodMatch);
        handleMethods.forEach(method -> {
            try {
                //反射Method转换为MethodHandle,提高效率
                MethodHandle handleMethod = lookup.findVirtual(handlerType, method.getName(), MethodType.methodType(method.getReturnType(), method.getParameterTypes()));
                TKey key = getKey(method);
                List<THandlerProxyInterface> handlers = handlerDict.computeIfAbsent(key, k -> new ArrayList<>());
                IObjectContainer objectContainer = getObjectContainer();
                if (objectContainer == null) {
                    throw new IllegalArgumentException("IObjectContainer is null");
                }
                // prototype
                THandlerProxyInterface handlerProxy = objectContainer.resolve(getHandlerProxyImplementationType());
                if (handlerProxy == null) {
                    throw new EnodeRuntimeException("THandlerProxyInterface is null, " + getHandlerProxyImplementationType().getName());
                }
                handlerProxy.setHandlerType(handlerType);
                handlerProxy.setMethod(method);
                handlerProxy.setMethodHandle(handleMethod);
                handlers.add(handlerProxy);
            } catch (Exception e) {
                throw new RegisterComponentException(e);
            }
        });
    }
}

File: enode/src/main/java/org/enodeframework/infrastructure/impl/AbstractHandlerProvider.kt
package org.enodeframework.infrastructure.impl

import org.enodeframework.annotation.Command
import org.enodeframework.annotation.Event
import org.enodeframework.annotation.Priority
import org.enodeframework.annotation.Subscribe
import org.enodeframework.common.container.ObjectContainer
import org.enodeframework.common.exception.HandlerRegisterException
import org.enodeframework.infrastructure.AssemblyInitializer
import org.enodeframework.infrastructure.MethodInvocation
import org.enodeframework.infrastructure.ObjectProxy
import org.enodeframework.messaging.MessageHandlerData
import org.reflections.ReflectionUtils
import org.slf4j.LoggerFactory
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import kotlin.reflect.jvm.kotlinFunction

abstract class AbstractHandlerProvider<TKey, THandlerProxyInterface, THandlerSource> :
    AssemblyInitializer where THandlerProxyInterface : ObjectProxy, THandlerProxyInterface : MethodInvocation {
    private val handlerDict: MutableMap<TKey, MutableList<THandlerProxyInterface>> = HashMap()
    private val messageHandlerDict: MutableMap<TKey, MessageHandlerData<THandlerProxyInterface>> = HashMap()
    private val lookup = MethodHandles.lookup()

    private val logger = LoggerFactory.getLogger(this.javaClass)

    /**
     * handlerDict中注册的key
     */
    protected abstract fun getKey(method: Method): TKey

    /**
     * 代理执行的类型
     */
    protected abstract fun getHandlerProxyImplementationType(): Class<out THandlerProxyInterface>

    /**
     * 判定执行类型是否匹配
     */
    protected abstract fun isHandlerSourceMatchKey(handlerSource: THandlerSource, key: TKey): Boolean

    /**
     * kotlin suspend 方法会多一个参数 kotlin.coroutines.Continuation
     */
    protected abstract fun isHandleMethodMatch(method: Method): Boolean

    /**
     * 在启动时校验一个消息是否可以注册多个处理器
     */
    protected abstract fun isHandleRegisterOnce(): Boolean

    /**
     * 是否是一个可挂起的方法
     */
    protected open fun isSuspendMethod(method: Method): Boolean {
        return method.kotlinFunction?.isSuspend == true
    }

    override fun initialize(objectContainer: ObjectContainer, componentTypes: Set<Class<*>>) {
        componentTypes.filter { type: Class<*> -> isHandlerType(type) }
            .forEach { handlerType: Class<*> -> registerHandler(objectContainer, handlerType) }
        initializeHandlerPriority()
    }

    fun getHandlersInternal(source: THandlerSource): List<MessageHandlerData<THandlerProxyInterface>> {
        return messageHandlerDict.keys.filter { key: TKey -> isHandlerSourceMatchKey(source, key) }
            .map { key: TKey -> messageHandlerDict.getValue(key) }
    }

    private fun initializeHandlerPriority() {
        handlerDict.forEach { (key: TKey, handlers: List<THandlerProxyInterface>) ->
            val handlerData = MessageHandlerData<THandlerProxyInterface>()
            val listHandlers: MutableList<THandlerProxyInterface> = ArrayList()
            val queueHandlerDict: MutableMap<THandlerProxyInterface, Int> = HashMap()
            handlers.forEach { handler: THandlerProxyInterface ->
                val priority = getHandleMethodPriority(handler)
                if (priority == 0) {
                    listHandlers.add(handler)
                } else {
                    queueHandlerDict[handler] = priority
                }
            }
            handlerData.allHandlers = handlers
            handlerData.listHandlers = listHandlers
            handlerData.queuedHandlers =
                queueHandlerDict.entries.sortedWith(Comparator.comparingInt { v -> v.value }).map { x -> x.key }
            messageHandlerDict[key] = handlerData
        }
    }

    private fun getHandleMethodPriority(handler: THandlerProxyInterface): Int {
        val method = handler.getMethod()
        var priority = 0
        val methodPriority = method.getAnnotation(Priority::class.java)
        if (methodPriority != null) {
            priority = methodPriority.value
        }
        if (priority == 0) {
            val classPriority = handler.getInnerObject().javaClass.getAnnotation(Priority::class.java)
            if (classPriority != null) {
                priority = classPriority.value
            }
        }
        return priority
    }

    private fun isHandlerType(type: Class<*>): Boolean {
        if (type.isInterface) {
            return false
        }
        return if (Modifier.isAbstract(type.modifiers)) {
            false
        } else containsHandleType(type)
    }

    private fun containsHandleType(type: Class<*>): Boolean {
        return type.isAnnotationPresent(Command::class.java) || type.isAnnotationPresent(Event::class.java)
    }

    protected fun isMethodAnnotationSubscribe(method: Method): Boolean {
        return method.isAnnotationPresent(Subscribe::class.java)
    }

    private fun registerHandler(objectContainer: ObjectContainer, handlerType: Class<*>) {
        val handleMethods = ReflectionUtils.getMethods(handlerType, { method: Method -> isHandleMethodMatch(method) })
        handleMethods.forEach { method: Method ->
            // 反射Method转换为MethodHandle，提高效率
            val handleMethod = lookup.findVirtual(
                handlerType, method.name, MethodType.methodType(method.returnType, method.parameterTypes)
            )
            val key = this.getKey(method)
            val handlers = handlerDict.computeIfAbsent(key) { ArrayList() }
            val handlerProxy = this.getHandlerProxyImplementationType().getDeclaredConstructor().newInstance()
            handlerProxy.setInnerObject(objectContainer.resolve(handlerType))
            handlerProxy.setMethod(method)
            handlerProxy.setMethodHandle(handleMethod)
            handlers.add(handlerProxy)
            // 针对command，只允许一个处理器
            if (isHandleRegisterOnce() && handlers.size > 1) {
                throw HandlerRegisterException("${method.name}#${method.parameterTypes.joinToString(",")}")
            }
            logger.info(
                "Handler register success. {} => {}",
                key,
                "${handlerType.name}#${method.name}#${method.parameterTypes.joinToString(",")}"
            )
        }
    }
}